### ISSUE TITLE
Consider moving away from JAXB, to be Android friendly

### ISSUE THREAD

**Issue by tomfitzhenry** [cternes/openkeepass#29]
I'm writing an Android app that uses openkeepass. When I read/write a DB, I see:
```
java.lang.ClassNotFoundException: Didn't find class "javax.xml.bind.JAXB" on path: DexPathList[[zip file "/data/app/tf.tom.myapplication-1/base.apk"],nativeLibraryDirectories=[/data/app/tf.tom.myapplication-1/lib/x86, /vendor/lib, /system/lib]]
    at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
    at de.slackspace.openkeepass.xml.KeePassDatabaseXmlParser.toXml(KeePassDatabaseXmlParser.java:18) 
    at de.slackspace.openkeepass.KeePassDatabase.write(KeePassDatabase.java:426) 
```

The reason is that Android's javax.xml does provide the entirety of JAXB, nor does Android allow apps to include JARs that are in javax.*

This was news to me, and pretty disappointing.

Consequently, if libraries wish to be Android friendly, they should choose XML libraries other than JAXB, such as Jackson.

Would you be willing to do this? I can provide a PR, when I'm not busy.

**Comment by cternes** [comment]
It's a shame that Android does not provide JAXB. That will be a really radical change since all of the XML serialization/deserialization has to be rewritten.

However, I will look into it to find out what options we have.

**Comment by cternes** [comment]
@tomfitzhenry I've tried Simple XML and I'm really disappointed, it does not work out as expected. 

Currently I'm looking at Jackson which looks much better but there is still a lot of work to do. I want to be sure that Jackson works on Android but I'm missing an Android device, could you check if Jackson works on your Android with the following dependencies:

    <dependency>
            <groupId>com.fasterxml.jackson.dataformat</groupId>
            <artifactId>jackson-dataformat-xml</artifactId>
            <version>2.6.3</version>
        </dependency>
        <dependency>
          <groupId>org.codehaus.woodstox</groupId>
          <artifactId>woodstox-core-asl</artifactId>
          <version>4.4.1</version>
        </dependency>
        <dependency>
            <groupId>javax.xml.stream</groupId>
            <artifactId>stax-api</artifactId>
            <version>1.0-2</version>
            <scope>compile</scope>
        </dependency>

**Comment by cternes** [comment]
I've created a branch 'feature-xml-jackson'. Could you check if that works on Android?

**Comment by vhschlenker** [comment]
I tried today to use openkeepass with your feature-xml-jackson.
Everything looked good, except one thing at the very end:
`com.android.build.api.transform.TransformException: com.android.builder.packaging.DuplicateFileException: Duplicate files copied in APK META-INF/services/com.fasterxml.jackson.core.ObjectCodec`
This error is known [using-jackson-dataformat-xml-on-android](http://stackoverflow.com/questions/31360025/using-jackson-dataformat-xml-on-android) but there does not seem to be a proper solution for it.
Looking at [android-keepass
](https://github.com/kreutzwj/android-keepass/blob/master/app/src/main/java/com/joelkreutzwieser/apps/keepass/keepass/KeePassDatabase.java#L33) Simple XML could work, as he used openkeepass successfully

**Comment by cternes** [comment]
Yeah, it is really a pitty that XML parsing is so hard on android. Maybe I will give SimpleXML a try if I have some spare time.

**Comment by vhschlenker** [comment]
I'd love to see that, since your lib is by far the best documented!
It seems like many other keepass projects on android use the [xmlpull](https://developer.android.com/reference/org/xmlpull/v1/package-summary.html) library.
For example keepassdroid use it for [reading](https://github.com/bpellin/keepassdroid/blob/master/app/src/main/java/com/keepassdroid/database/load/ImporterV4.java) and [writing](https://github.com/bpellin/keepassdroid/blob/master/app/src/main/java/com/keepassdroid/database/save/PwDbV4Output.java) and here is another usage example in a [javalib](https://github.com/pfn/keepassj/blob/master/src/main/java/com/hanhuy/keepassj/KdbxFile.java).

**Comment by cternes** [comment]
@vhschlenker I've managed to switch to SimpleXML on the new branch 'feature-simplexml'. Could you help me and try if this works better on Android?

**Comment by vhschlenker** [comment]
Venit, vidit, vicit! (It's the third person form of the well known quote :smile: )
I just needed to change a few things to avoid some Android quirks (like the bouncycastle <-> spongycastle thing) and a gradle file to build it as gradle module and voila, it works!
(You can view my changes [here](https://gitlab.com/vhschlenker/openkeepass/commit/45c4a5c7331bf5cde715ce2737b49f9e88eb29d5)
If you want, I can break down the commit and make a merge request with only the android parts etc. .
And now I can start working on finalising the app :+1: .
Just one teeny tiny thing,  three of the tests are failing with the equalsverifier because of a recursive datastructure:
 `Add prefab values for one of the following types: History, List<Entry>, Entry.
at de.slackspace.openkeepass.domain.HistoryTest.equalsContract(HistoryTest.java:12)
Add prefab values for one of the following types: Entry, History, List<Entry>.
at de.slackspace.openkeepass.domain.EntryTest.equalsContract(EntryTest.java:20)
Add prefab values for one of the following types: Group, List<Entry>, Entry, History.
at de.slackspace.openkeepass.domain.GroupTest.equalsContract(GroupTest.java:20)`
can you confirm this?

**Comment by cternes** [comment]
Thanks for testing! I will look at the changes you did and will try to apply them. 

About the test, I'm wondering because all the tests are running fine locally on my machine as well as on the Travis CI.

**Comment by cternes** [comment]
There was another quirk with SpongyCastle on other platforms than Android. I've removed the security provider stuff in order to make it work. It should now work on Android out of the box as well as on other platforms.

**Comment by vhschlenker** [comment]
Your changes work as expected :+1: .
Regarding the failing test: I used a newer version of the equalsverifier (2.x.x), where the tests fail. With the version in the pom (1.7.7) the tests passes.
The only thing wich might be still problematic with it could be this:
`configurations {
    compile.exclude module: 'stax'
    compile.exclude module: 'stax-api'
    compile.exclude module: 'xpp3'
}`
Which is needed to avoid this `Error:Ill-advised or mistaken usage of a core class (java.* or javax.*)...` .
But I did not tried yet to import it as a precompiled jar, since Android Studio does not support macen/pom builds as modules.
